### PR TITLE
WIP: DEV-68 Integracja ze stanem banków krwi (krew.info)

### PR DESCRIPTION
**Pull request by adjurk**
**Implements:**

- JSON output with valid schema, adhering to the agreed format for backend.

**TODO:**

- [x] Refactor to a `main()` function

### CODE REVIEW

File: scraper.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+url = 'https://krew.info/zapasy/'
+
+response = requests.get(url)
+# print(response.status_code)
+response.encoding = 'utf-8'
+
+soup = BeautifulSoup(response.text, 'html.parser')
+table = soup.find('table')
+# print(table)
```
Comment (s-kakol) [inline scraper.py]: Debug prints should be removed imho

File: scraper.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+"""
+0 = stop
+11 = full (but not halted)
+1 = almost full
+2 = half
+3 = very low
```
Comment (s-kakol) [inline scraper.py]: Outdated and can be removed I believe

File: scraper.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    # print(imgs)
+    for img in imgs:
+        bank_status.append(map_img_to_blood_type(img.get('src')))
+    # print(bank_status)
+    return bank_status
+
+
+def get_datetime_modified(soup):
+    datetime_full = soup.find(string=re.compile("Aktualizacja stanu:")) # most reliable way for now, since this tag has no id
+    return re.search(r'\d.+', datetime_full).group(0)
```
Comment (s-kakol) [inline scraper.py]: Is the datetime returned a ISO8601 date? If the output is similar to `"02.04.2023 12:30"` from `README.md` then this needs to be fixed.

File: scraper.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    response = requests.get(url)
+    response.encoding = 'utf-8'
+
+    soup = BeautifulSoup(response.text, 'html.parser')
+    table = soup.find('table')
+
+    for blood_banks in table.find_all('tbody'):
+        rows = blood_banks.find_all('tr')
+
+    blood_types = get_all_blood_types(rows)
```
Comment (s-kakol) [inline scraper.py]: `rows` are potentially undefined

File: scraper.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    return blood_types
+
+
+# ['Białystok', 'Bydgoszcz', 'Gdańsk', 'Kalisz', 'Katowice', 'Kielce', 'Kraków', 'Lublin', 'Łódź', 'Olsztyn', 'Opole', 'Poznań', 'Racibórz', 'Radom', 'Rzeszów', 'Słupsk', 'Szczecin', 'Wałbrzych', 'Warszawa', 'Wrocław', 'Zielona Góra']
+def get_all_cities(rows):
+    cities = []
+    for row in rows:
+        imgs = row.find_all('img')
+    for img in imgs:
+        cities.append(img.get('alt'))
```
Comment (s-kakol) [inline scraper.py:22]: `imgs` are potentially undefined.

File: scraper.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    soup = BeautifulSoup(response.text, 'html.parser')
+    table = soup.find('table')
+
+    for blood_banks in table.find_all('tbody'):
+        rows = blood_banks.find_all('tr')
+
+    blood_types = get_all_blood_types(rows)
+    cities = get_all_cities(rows)
+
+    bank_status = []
```
Comment (s-kakol) [inline scraper.py]: Nitpicky but imho cleaner if variables are declared at the top of the function (look: `bank_status`, `blood_banks`). Do as you wish.

File: scraper.py [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    output_json = {
+        "datetime_modified": get_datetime_modified(soup),
+        "url_src": response.url,
+        "blood_banks": blood_banks
+    }
+
+    for count_cities, city in enumerate(cities):
+        blood_banks[city] = {}
+        for count_bd, blood_type in enumerate(blood_types):
+            blood_banks[city][blood_type] = bank_status[count_bd][count_cities]
```
Comment (s-kakol) [inline scraper.py]: First we prepare the response and then we modify it again. At least change the sequence.

**Review by s-kakol — APPROVED** [review]
Overall looks good to me, can be merged but added some comments for qol.

File: scraper.py (hunk 3 above)
Comment (adjurk) [reply scraper.py]: It's now in the desired format, thanks!

File: scraper.py (hunk 6 above)
Comment (adjurk) [reply scraper.py]: While with `bank_status` I can fully agree (and just moved this list to the top of the `main()` function), I cannot move `blood_banks` dict outside the for each loop as it'd be outside of scope of the `city` variable.

Comment (s-kakol) [reply scraper.py]: Right, I see that now. Ok Thx!

File: scraper.py (hunk 7 above)
Comment (adjurk) [reply scraper.py]: Hmm, you're right that it looks confusing in terms of when the loop is called and what it refers to.

I've swapped the order of the `output_json` and the for-each loop, thanks!

**Comment by s-kakol** [comment]
LGTM!

**Pull request merged**